File: python/owrt_monitor/build_log.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

# enw-device-firmware top-level Makefile prints this on a successful build.
# Example: ">>>> owrt2102.asus_mt_wifi7_mt7987  Build done in: 05:07.537"
_SUCCESS_DONE = re.compile(r"^>>>> (?P<target>\S+)\s+Build done in:\s+(?P<duration>[\d:\.]+)\s*$")

# `No space left on device` is high-signal: if it appears anywhere in the log we
# call this disk_full regardless of any subsequent generic "Error 2" lines, since
# downstream errors are consequences of the disk being full.
_DISK_FULL = re.compile(r"No space left on device", re.IGNORECASE)

# Final make target that failed at the top level — the profile name itself.
# Example: "make: *** [include/owrt2102.mk:163: owrt2102.asus_mt_wifi7_mt7987] Error 2"
_TOPLEVEL_FAIL = re.compile(
    r"^make:\s+\*\*\*\s+\[(?P<makefile>[^\]]+):\s*(?P<target>\S+)\]\s+Error\s+\d+\s*$"
)

# A package or subdirectory that failed during the build, e.g.
# "make[3]: *** [package/feeds/mtk/foo/compile] Error 2"
_PACKAGE_FAIL = re.compile(
    r"^make\[\d+\]:\s+\*\*\*\s+\[(?P<step>[^\]]+)\]\s+Error\s+\d+\s*$"
)

# Non-fatal warnings emitted by OpenWrt's package metadata pass.
_WARNING = re.compile(r"^WARNING:\s+(?P<message>.*)$")

# Noise lines we should ignore when picking evidence (they almost always appear
# because the firmware tree calls `git describe` for versioning):
_KNOWN_NOISE_PATTERNS = (
    re.compile(r"^fatal: No names found, cannot describe anything\.\s*$"),
    re.compile(r"^cat: write error: No space left on device\s*$"),  # not the canonical signal
)

_MAX_WARNINGS = 50
_MAX_EVIDENCE_LINES = 5


@dataclass(frozen=True)
class BuildLogSummary:
    classification: str
    success: bool
    duration_sec: float | None = None
    failed_target: str | None = None
    failed_step: str | None = None
    failed_package: str | None = None
    evidence: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, object]:
        return {
            "classification": self.classification,
            "success": self.success,
            "duration_sec": self.duration_sec,
            "failed_target": self.failed_target,
            "failed_step": self.failed_step,
            "failed_package": self.failed_package,
            "evidence": list(self.evidence),
            "warnings": list(self.warnings),
        }


def _extract_failed_package(step: str | None) -> str | None:
    """Derive a human-readable package name from a make step path.

    Examples:
      package/foo/bar/compile          → foo/bar
      package/feeds/mtk/flowtable/install → feeds/mtk/flowtable
      target/linux/install             → target/linux
      /abs/path/...mk:228: world       → None  (synthetic toplevel-mk target, not a package)
    """
    if not step:
        return None
    # Synthetic make targets that aren't real packages.
    if step.endswith("world") or "toplevel.mk" in step:
        return None
    # Strip the trailing make target (compile/install/configure/etc).
    parts = step.split("/")
    if len(parts) < 2:
        return None
    leaf_targets = {"compile", "install", "configure"}
    if parts[0] == "package":
        package_parts = parts[1:-1] if parts[-1] in leaf_targets else parts[1:]
        return "/".join(package_parts) if package_parts else None
    if parts[0] == "target":
        package_parts = parts[1:-1] if parts[-1] in leaf_targets else parts[1:]
        return "target/" + "/".join(package_parts) if package_parts else None
    return None


def classify_build_log(log_path: Path) -> BuildLogSummary:
    """Classify a build.log into a structured summary.

    Reads the entire file (build logs for this lab top out at ~tens of KB).
    """
    if not log_path.exists():
        return BuildLogSummary(classification="missing_log", success=False)

    try:
        lines = log_path.read_text(encoding="utf-8", errors="replace").splitlines()
    except OSError as exc:
        return BuildLogSummary(
            classification="unreadable_log",
            success=False,
            evidence=[str(exc)],
        )

    success_match: re.Match[str] | None = None
    toplevel_fail: re.Match[str] | None = None
    package_fail: re.Match[str] | None = None
    disk_full_lines: list[str] = []
    warnings: list[str] = []

    for line in lines:
        m_success = _SUCCESS_DONE.match(line)
        if m_success is not None:
            success_match = m_success
            continue

        if _DISK_FULL.search(line) and not any(p.match(line) for p in _KNOWN_NOISE_PATTERNS):
            disk_full_lines.append(line)
            continue

        m_top = _TOPLEVEL_FAIL.match(line)
        if m_top is not None:
            toplevel_fail = m_top
            continue

        m_pkg = _PACKAGE_FAIL.match(line)
        if m_pkg is not None:
            # Keep the FIRST match: in make output the deepest failure is reported
            # first, with outer make levels appending propagation lines. The first
            # `make[N]: *** [...] Error N` is therefore the actual root cause.
            if package_fail is None:
                package_fail = m_pkg
            continue

        m_warn = _WARNING.match(line)
        if m_warn is not None and len(warnings) < _MAX_WARNINGS:
            warnings.append(m_warn.group("message"))

    if success_match is not None:
        duration = _parse_duration(success_match.group("duration"))
        return BuildLogSummary(
            classification="success",
            success=True,
            duration_sec=duration,
            failed_target=None,
            failed_step=None,
            evidence=[success_match.group(0)],
            warnings=warnings,
        )

    if disk_full_lines:
        evidence = disk_full_lines[:_MAX_EVIDENCE_LINES]
        return BuildLogSummary(
            classification="disk_full",
            success=False,
            failed_target=toplevel_fail.group("target") if toplevel_fail else None,
            evidence=evidence,
            warnings=warnings,
        )

    if package_fail is not None:
        step = package_fail.group("step")
        return BuildLogSummary(
            classification="failed_package",
            success=False,
            failed_target=toplevel_fail.group("target") if toplevel_fail else None,
            failed_step=step,
            failed_package=_extract_failed_package(step),
            evidence=[package_fail.group(0)],
            warnings=warnings,
        )

    if toplevel_fail is not None:
        return BuildLogSummary(
            classification="compile_error",
            success=False,
            failed_target=toplevel_fail.group("target"),
            evidence=[toplevel_fail.group(0)],
            warnings=warnings,
        )

    # No success marker, no recognised failure marker. Could be a partial log
    # (process killed mid-stream) or a wholly unrecognised failure shape.
    tail_evidence = [
        line for line in lines[-_MAX_EVIDENCE_LINES:] if line.strip()
    ]
    return BuildLogSummary(
        classification="unknown",
        success=False,
        evidence=tail_evidence,
        warnings=warnings,
    )


def _parse_duration(value: str) -> float | None:
    """Parse 'MM:SS.fff' or 'HH:MM:SS.fff' into seconds."""
    parts = value.split(":")
    try:
        if len(parts) == 2:
            minutes, seconds = parts
            return int(minutes) * 60 + float(seconds)
        if len(parts) == 3:
            hours, minutes, seconds = parts
            return int(hours) * 3600 + int(minutes) * 60 + float(seconds)
    except ValueError:
        return None
    return None

File: internal/buildlog/buildlog.go
// Package buildlog is the Go port of python/owrt_monitor/build_log.py: it
// classifies an OpenWrt build.log into a structured summary (success /
// disk_full / failed_package / compile_error / unknown) with evidence,
// warnings, and a parsed duration. The summary shape matches BuildLogSummary
// .to_dict so reports are identical across engines.
package buildlog

import (
	"os"
	"regexp"
	"strconv"
	"strings"
)

var (
	successDone  = regexp.MustCompile(`^>>>> (?P<target>\S+)\s+Build done in:\s+(?P<duration>[\d:.]+)\s*$`)
	diskFull     = regexp.MustCompile(`(?i)No space left on device`)
	toplevelFail = regexp.MustCompile(`^make:\s+\*\*\*\s+\[(?P<makefile>[^\]]+):\s*(?P<target>\S+)\]\s+Error\s+\d+\s*$`)
	packageFail  = regexp.MustCompile(`^make\[\d+\]:\s+\*\*\*\s+\[(?P<step>[^\]]+)\]\s+Error\s+\d+\s*$`)
	warningRe    = regexp.MustCompile(`^WARNING:\s+(?P<message>.*)$`)
	knownNoise   = []*regexp.Regexp{
		regexp.MustCompile(`^fatal: No names found, cannot describe anything\.\s*$`),
		regexp.MustCompile(`^cat: write error: No space left on device\s*$`),
	}
)

const (
	maxWarnings     = 50
	maxEvidenceLine = 5
)

// Summary mirrors build_log.BuildLogSummary.
type Summary struct {
	Classification string
	Success        bool
	DurationSec    *float64
	FailedTarget   string
	FailedStep     string
	FailedPackage  string
	Evidence       []string
	Warnings       []string
}

// ToMap renders the summary into BuildLogSummary.to_dict's shape (nullable
// fields become null, list fields default to []).
func (s Summary) ToMap() map[string]any {
	return map[string]any{
		"classification": s.Classification,
		"success":        s.Success,
		"duration_sec":   floatOrNil(s.DurationSec),
		"failed_target":  strOrNil(s.FailedTarget),
		"failed_step":    strOrNil(s.FailedStep),
		"failed_package": strOrNil(s.FailedPackage),
		"evidence":       listOrEmpty(s.Evidence),
		"warnings":       listOrEmpty(s.Warnings),
	}
}

// Classify reads and classifies a build.log (build_log.classify_build_log).
func Classify(logPath string) Summary {
	data, err := os.ReadFile(logPath)
	if err != nil {
		if os.IsNotExist(err) {
			return Summary{Classification: "missing_log", Success: false}
		}
		return Summary{Classification: "unreadable_log", Success: false, Evidence: []string{err.Error()}}
	}
	lines := strings.Split(strings.ReplaceAll(string(data), "\r\n", "\n"), "\n")
	// splitlines() drops a trailing empty element from a final newline.
	if n := len(lines); n > 0 && lines[n-1] == "" {
		lines = lines[:n-1]
	}

	var (
		successM, toplevelM, packageM []string
		diskFullLines                 []string
		warnings                      []string
	)

	for _, line := range lines {
		if m := successDone.FindStringSubmatch(line); m != nil {
			successM = m
			continue
		}
		if diskFull.MatchString(line) && !matchesAnyNoise(line) {
			diskFullLines = append(diskFullLines, line)
			continue
		}
		if m := toplevelFail.FindStringSubmatch(line); m != nil {
			toplevelM = m
			continue
		}
		if m := packageFail.FindStringSubmatch(line); m != nil {
			if packageM == nil { // keep the first (root-cause) match
				packageM = m
			}
			continue
		}
		if m := warningRe.FindStringSubmatch(line); m != nil && len(warnings) < maxWarnings {
			warnings = append(warnings, group(warningRe, m, "message"))
		}
	}

	toplevelTarget := ""
	if toplevelM != nil {
		toplevelTarget = group(toplevelFail, toplevelM, "target")
	}

	switch {
	case successM != nil:
		return Summary{
			Classification: "success", Success: true,
			DurationSec: parseDuration(group(successDone, successM, "duration")),
			Evidence:    []string{successM[0]}, Warnings: warnings,
		}
	case len(diskFullLines) > 0:
		return Summary{
			Classification: "disk_full", Success: false, FailedTarget: toplevelTarget,
			Evidence: head(diskFullLines, maxEvidenceLine), Warnings: warnings,
		}
	case packageM != nil:
		step := group(packageFail, packageM, "step")
		return Summary{
			Classification: "failed_package", Success: false, FailedTarget: toplevelTarget,
			FailedStep: step, FailedPackage: extractFailedPackage(step),
			Evidence: []string{packageM[0]}, Warnings: warnings,
		}
	case toplevelM != nil:
		return Summary{
			Classification: "compile_error", Success: false, FailedTarget: toplevelTarget,
			Evidence: []string{toplevelM[0]}, Warnings: warnings,
		}
	default:
		return Summary{
			Classification: "unknown", Success: false,
			Evidence: nonBlankTail(lines, maxEvidenceLine), Warnings: warnings,
		}
	}
}

func extractFailedPackage(step string) string {
	if step == "" {
		return ""
	}
	if strings.HasSuffix(step, "world") || strings.Contains(step, "toplevel.mk") {
		return ""
	}
	parts := strings.Split(step, "/")
	if len(parts) < 2 {
		return ""
	}
	leaf := map[string]bool{"compile": true, "install": true, "configure": true}
	switch parts[0] {
	case "package":
		pp := parts[1:]
		if leaf[parts[len(parts)-1]] {
			pp = parts[1 : len(parts)-1]
		}
		if len(pp) == 0 {
			return ""
		}
		return strings.Join(pp, "/")
	case "target":
		pp := parts[1:]
		if leaf[parts[len(parts)-1]] {
			pp = parts[1 : len(parts)-1]
		}
		if len(pp) == 0 {
			return ""
		}
		return "target/" + strings.Join(pp, "/")
	}
	return ""
}

func parseDuration(value string) *float64 {
	parts := strings.Split(value, ":")
	switch len(parts) {
	case 2:
		m, err1 := strconv.Atoi(parts[0])
		s, err2 := strconv.ParseFloat(parts[1], 64)
		if err1 != nil || err2 != nil {
			return nil
		}
		return f(float64(m)*60 + s)
	case 3:
		h, err1 := strconv.Atoi(parts[0])
		m, err2 := strconv.Atoi(parts[1])
		s, err3 := strconv.ParseFloat(parts[2], 64)
		if err1 != nil || err2 != nil || err3 != nil {
			return nil
		}
		return f(float64(h)*3600 + float64(m)*60 + s)
	}
	return nil
}

func matchesAnyNoise(line string) bool {
	for _, re := range knownNoise {
		if re.MatchString(line) {
			return true
		}
	}
	return false
}

func group(re *regexp.Regexp, match []string, name string) string {
	idx := re.SubexpIndex(name)
	if idx < 0 || idx >= len(match) {
		return ""
	}
	return match[idx]
}

func head(s []string, n int) []string {
	if len(s) > n {
		return append([]string{}, s[:n]...)
	}
	return append([]string{}, s...)
}

func nonBlankTail(lines []string, n int) []string {
	start := 0
	if len(lines) > n {
		start = len(lines) - n
	}
	out := []string{}
	for _, line := range lines[start:] {
		if strings.TrimSpace(line) != "" {
			out = append(out, line)
		}
	}
	return out
}

func f(v float64) *float64 { return &v }

func floatOrNil(p *float64) any {
	if p == nil {
		return nil
	}
	return *p
}

func strOrNil(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func listOrEmpty(s []string) []string {
	if s == nil {
		return []string{}
	}
	return s
}
